File: framework/ndn_iot_python/ndn_iot_python/bootstrap/bootstrap.py
#!/usr/bin/python

from pyndn import Name, Interest, Data, Exclude

from pyndn.threadsafe_face import ThreadsafeFace
from pyndn.security import KeyChain, SecurityException
from pyndn.security.identity import IdentityManager, FilePrivateKeyStorage, BasicIdentityStorage
from pyndn.security.policy import ConfigPolicyManager
from pyndn.security.certificate import IdentityCertificate

from pyndn.encoding import ProtobufTlv
from pyndn.util import MemoryContentCache
from pyndn.util.boost_info_parser import BoostInfoParser

from ndn_iot_python.commands.app_request_pb2 import AppRequestMessage

import time
import os, sys
import json

import logging

try:
    import asyncio
except ImportError:
    import trollius as asyncio

class Bootstrap(object):
    """
    Create a Bootstrap object. Bootstrap object provides interface for setting up KeyChain, default certificate name;
    (as a producer) requesting publishing authorization from controller; and (as a consumer) keeping track of changes

    :param face: the face for communicating with a local / remote forwarder
    :type face: ThreadsafeFace

    TODO: support Face as well as ThreadsafeFace
    """
    def __init__(self, face):
        self._defaultIdentity = None
        self._defaultCertificateName = None
        
        self._controllerName = None
        self._controllerCertificate = None

        self._applicationName = ""

        self._identityManager = IdentityManager(BasicIdentityStorage(), FilePrivateKeyStorage())
        __location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))

        self._policyManager = ConfigPolicyManager(os.path.join(__location__, ".default.conf"))
        # keyChain is what we return to the application after successful setup
        # TODO: should we separate keyChain from internal KeyChain used to verify trust schemas?
        self._keyChain = KeyChain(self._identityManager, self._policyManager)

        self._face = face
        # setFace for keyChain or else it won't be able to express interests for certs
        self._keyChain.setFace(self._face)
        self._certificateContentCache = MemoryContentCache(face)
        
        self._trustSchemas = dict()

###############################################
# Initial keyChain and defaultCertificate setup
###############################################
    def setupDefaultIdentityAndRoot(self, defaultIdentityOrFileName, signerName = None, onSetupComplete = None, onSetupFailed = None):
        """
        Sets up the keyChain, default key name and certificate name according to given 
        configuration. If successful, this KeyChain and default certificate name will be 
        returned to the application, which can be passed to instances like Consumer, Discovery, etc

        :param defaultIdentityOrFileName: if str, the name of the configuration file; if Name, 
          the default identity name of this IoT node. The node will use the default keys and 
          certificate of that identity name.
        :type defaultIdentityOrFileName: Name or str
        :param signerName: (optional) the expected signing identity of the certificate
        :type signerName: Name
        :param onSetupComplete: (optional) onSetupComplete(Name, KeyChain) will be called if 
          set up's successful
        :type onSetupComplete: function object
        :param onSetupFailed: (optional) onSetupFailed(msg) will be called if setup fails
        :type onSetupFailed: function object
        """
        def helper(identityName, signerName):
            try:
                self._defaultIdentity = identityName
                self._defaultCertificateName = self._identityManager.getDefaultCertificateNameForIdentity(self._defaultIdentity)
                self._defaultKeyName = self._identityManager.getDefaultKeyNameForIdentity(identityName)
            except SecurityException:
                msg = "Identity " + identityName.toUri() + " in configuration does not exist. Please configure the device with this identity first"
                print msg
                if onSetupFailed:
                    onSetupFailed(msg)
            
            # Note we'll not be able to issue face commands before this point
            self._face.setCommandSigningInfo(self._keyChain, self._defaultCertificateName)
            print "default cert name " + self._defaultCertificateName.toUri()
            # Serve our own certificate
            self._certificateContentCache.registerPrefix(Name(self._defaultCertificateName).getPrefix(-1), self.onRegisterFailed)
            self._certificateContentCache.add(self._keyChain.getCertificate(self._defaultCertificateName))

            actualSignerName = self._keyChain.getCertificate(self._defaultCertificateName).getSignature().getKeyLocator().getKeyName()    
            if not signerName:
                print "Deriving from " + actualSignerName.toUri() + " for controller name"
            else:
                if signerName and actualSignerName.toUri() != signerName.toUri():
                    msg = "Configuration signer names mismatch: expected " + signerName.toUri() + "; got " + actualSignerName.toUri()
                    print msg
                    if onSetupFailed:
                        onSetupFailed(msg)

            self._controllerName = self.getIdentityNameFromCertName(actualSignerName)
            print "Controller name: " + self._controllerName.toUri()

            try:
                self._controllerCertificate = self._keyChain.getCertificate(self._identityManager.getDefaultCertificateNameForIdentity(self._controllerName))
                
                # TODO: this does not seem a good approach, implementation-wise and security implication
                self._policyManager._certificateCache.insertCertificate(self._controllerCertificate)
                if onSetupComplete:
                    onSetupComplete(Name(self._defaultCertificateName), self._keyChain)
            except SecurityException as e:
                print "don't have controller certificate " + actualSignerName.toUri() + " yet"
                controllerCertInterest = Interest(Name(actualSignerName))
                controllerCertInterest.setInterestLifetimeMilliseconds(4000)
                
                self._face.expressInterest(controllerCertInterest, 
                  lambda interest, data: self.onControllerCertData(interest, data, onSetupComplete, onSetupFailed), 
                  lambda interest: self.onControllerCertTimeout(interest, onSetupComplete, onSetupFailed))
            return

        if isinstance(defaultIdentityOrFileName, basestring):
            confObj = self.processConfiguration(defaultIdentityOrFileName)
            if "identity" in confObj:
                if confObj["identity"] == "default":
                    # TODO: handling the case where no default identity is present
                    defaultIdentity = self._keyChain.getDefaultIdentity()
                else:
                    defaultIdentity = Name(confObj["identity"])
            else:
                defaultIdentity = self._keyChain.getDefaultIdentity()

            # TODO: handling signature with direct bits instead of keylocator keyname
            if "signer" in confObj:    
                if confObj["signer"] == "default":
                    signerName = None
                else:
                    signerName = Name(confObj["signer"])
            else:
                signerName = None
                print "Deriving from " + signerName.toUri() + " for controller name"

            helper(defaultIdentity, signerName)
        else:
            if isinstance(defaultIdentityOrFileName, Name):
                helper(defaultIdentityOrFileName, signerName)
            else:
                raise RuntimeError("Please call setupDefaultIdentityAndRoot with identity name and root key name")
        return

    def onControllerCertData(self, interest, data, onSetupComplete, onSetupFailed):
        # TODO: verification rule for received self-signed cert. 
        # So, if a controller comes masquerading in at this point with the right name, it is problematic. Similar with ndn-pi's implementation
        self._controllerCertificate = IdentityCertificate(data)
        # insert root certificate so that we could verify initial trust schemas
        # TODO: this does not seem a good approach, implementation-wise and security implication
        self._keyChain.getPolicyManager()._certificateCache.insertCertificate(self._controllerCertificate)
        try:
            self._identityManager.addCertificate(self._controllerCertificate)
        except SecurityException as e:
            print str(e)
        for schema in self._trustSchemas:
            # TODO: remove the concept of pending-schema
            if "pending-schema" in self._trustSchemas[schema]:
                self._keyChain.verifyData(self._trustSchemas[schema]["pending-schema"], self.onSchemaVerified, self.onSchemaVerificationFailed)
        if onSetupComplete:
            onSetupComplete(Name(self._defaultCertificateName), self._keyChain)
        return

    def onControllerCertTimeout(self, interest, onSetupComplete, onSetupFailed):
        print "Controller certificate interest times out"
        newInterest = Interest(interest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest, 
          lambda interest, data: self.onControllerCertData(interest, data, onSetupComplete, onSetupFailed), 
          lambda interest: self.onControllerCertTimeout(interest, onSetupComplete, onSetupFailed))
        return

#########################################################
# Handling application consumption (trust schema updates)
#########################################################
    # TODO: if trust schema gets over packet size limit, segmentation
    def startTrustSchemaUpdate(self, appPrefix, onUpdateSuccess = None, onUpdateFailed = None):
        """
        Starts trust schema update for under an application prefix: initial 
        interest asks for the rightMostChild, and later interests are sent 
        with previous version excluded. Each verified trust schema will trigger
        onUpdateSuccess and update the ConfigPolicyManager for the keyChain
        in this instance, and unverified ones will trigger onUpdateFailed.

        The keyChain and trust anchor should be set up using setupDefaultIdentityAndRoot
        before calling this method. 

        :param appPrefix: the prefix to ask trust schema for. (interest name: /<prefix>/_schema)
        :type appPrefix: Name
        :param onUpdateSuccess: (optional) onUpdateSuccess(trustSchemaStr, isInitial) is 
          called when update succeeds
        :type onUpdateSuccess: function object
        :param onUpdateFailed: (optional) onUpdateFailed(msg) is called when update fails
        :type onUpdateFailed: function object
        """
        namespace = appPrefix.toUri()
        if namespace in self._trustSchemas:
            if self._trustSchemas[namespace]["following"] == True:
                print "Already following trust schema under this namespace!"
                return
            self._trustSchemas[namespace]["following"] = True
        else:
            self._trustSchemas[namespace] = {"following": True, "version": 0, "is-initial": True}

        initialInterest = Interest(Name(namespace).append("_schema"))
        initialInterest.setChildSelector(1)
        self._face.expressInterest(initialInterest, 
          lambda interest, data: self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed), 
          lambda interest: self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed))
        return

    def stopTrustSchemaUpdate(self):
        print "stopTrustSchemaUpdate not implemented"
        return

    def onSchemaVerified(self, data, onUpdateSuccess, onUpdateFailed):
        print "trust schema verified: " + data.getName().toUri()
        version = data.getName().get(-1)
        namespace = data.getName().getPrefix(-2).toUri()
        if not (namespace in self._trustSchemas):
            print "unexpected: received trust schema for application namespace that's not being followed; malformed data name?"
            return

        if version.toVersion() <= self._trustSchemas[namespace]["version"]:
            msg = "Got out-of-date trust schema"
            print msg
            if onUpdateFailed:
                onUpdateFailed(msg)
            return

        self._trustSchemas[namespace]["version"] = version.toVersion()
        
        if "pending-schema" in self._trustSchemas[namespace] and self._trustSchemas[namespace]["pending-schema"].getName().toUri() == data.getName().toUri():
            # we verified a pending trust schema, don't need to keep that any more
            del self._trustSchemas[namespace]["pending-schema"]

        self._trustSchemas[namespace]["trust-schema"] = data.getContent().toRawStr()
        print self._trustSchemas[namespace]["trust-schema"]

        # TODO: what about trust schema for discovery, is discovery its own application?
        newInterest = Interest(Name(data.getName()).getPrefix(-1))
        newInterest.setChildSelector(1)
        exclude = Exclude()
        exclude.appendAny()
        exclude.appendComponent(version)
        newInterest.setExclude(exclude)
        self._face.expressInterest(newInterest, 
          lambda interest, data: self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed), 
          lambda interest: self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed))

        # Note: this changes the verification rules for root cert, future trust schemas as well; ideally from the outside this doesn't have an impact, but do we want to avoid this?
        # Per reset function in ConfigPolicyManager; For now we don't call reset as we still want root cert in our certCache, instead of asking for it again (when we want to verify) each time we update the trust schema
        self._policyManager.config = BoostInfoParser()
        self._policyManager.config.read(self._trustSchemas[namespace]["trust-schema"], "updated-schema")
        
        if onUpdateSuccess:
            onUpdateSuccess(data.getContent().toRawStr(), self._trustSchemas[namespace]["is-initial"])
        self._trustSchemas[namespace]["is-initial"] = False
        return

    def onSchemaVerificationFailed(self, data, reason, onUpdateSuccess, onUpdateFailed):
        print "trust schema verification failed: " + reason
        namespace = data.getName().getPrefix(-2).toUri()
        if not (namespace in self._trustSchemas):
            print "unexpected: received trust schema for application namespace that's not being followed; malformed data name?"
            return
        
        newInterest = Interest(Name(data.getName()).getPrefix(-1))
        newInterest.setChildSelector(1)
        exclude = Exclude()
        exclude.appendAny()
        exclude.appendComponent(Name.Component.fromVersion(self._trustSchemas[namespace]["version"]))
        newInterest.setExclude(exclude)
        # Don't immediately ask for potentially the same content again if verification fails
        self._face.callLater(4000, lambda : 
          self._face.expressInterest(newInterest, 
            lambda interest, data: self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed), 
            lambda interest: self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed)))
        return

    def onTrustSchemaData(self, interest, data, onUpdateSuccess, onUpdateFailed):
        print("Trust schema received: " + data.getName().toUri())
        namespace = data.getName().getPrefix(-2).toUri()
        # Process newly received trust schema
        if not self._controllerCertificate:
            # we don't yet have the root certificate fetched, so we store this cert for now
            print "Controller certificate not yet present, verify once it's in place"
            self._trustSchemas[namespace]["pending-schema"] = data
        else:
            # we veriy the received trust schema, should we use an internal KeyChain instead?
            self._keyChain.verifyData(data, 
              lambda data: self.onSchemaVerified(data, onUpdateSuccess, onUpdateFailed), 
              lambda data, reason: self.onSchemaVerificationFailed(data, reason, onUpdateSuccess, onUpdateFailed))

        return

    def onTrustSchemaTimeout(self, interest, onUpdateSuccess, onUpdateFailed):
        print("Trust schema interest times out: " + interest.getName().toUri())
        newInterest = Interest(interest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest, 
          lambda interest, data: self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed), 
          lambda interest: self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed))        
        return

###############################################
# Handling application producing authorizations
###############################################
    # Wrapper for sendAppRequest, fills in already configured defaultCertificateName
    def requestProducerAuthorization(self, dataPrefix, appName, onRequestSuccess = None, onRequestFailed = None):
        """
        Requests producing authorization for a data prefix: commandInterest is sent out 
        to the controller, using /<controller identity>/requests/<encoded-application-parameters>/<signed-interest-suffix>
        where encoded-application-parameters is a ProtobufTlv encoding of 
        {appPrefix, certificateName, appName}

        The keyChain, trust anchor and controller name should be set up using 
        setupDefaultIdentityAndRoot before calling this method.

        :param dataPrefix: the prefix to request publishing for
        :type dataPrefix: Name
        :param appName: the application name to request publishing for
        :type appName: str
        :param onRequestSuccess: (optional) onRequestSuccess() is called when a valid response
          if received for the request
        :type onRequestSuccess: function object
        :param onRequestFailed: (optional) onRequestFailed(msg) is called when request fails
        :type onRequestFailed: function object
        """
        # TODO: update logic on this part, should the presence of default certificate name be mandatory? 
        # And allow application developer to send app request to a configured root/controller?
        if not self._defaultCertificateName:
            raise RuntimeError("Default certificate is missing! Try setupDefaultIdentityAndRoot first?")
            return
        self.sendAppRequest(self._defaultCertificateName, dataPrefix, appName, onRequestSuccess, onRequestFailed)

    def sendAppRequest(self, certificateName, dataPrefix, applicationName, onRequestSuccess, onRequestFailed):
        message = AppRequestMessage()

        for component in range(certificateName.size()):
            message.command.idName.components.append(certificateName.get(component).toEscapedString())
        for component in range(dataPrefix.size()):
            message.command.dataPrefix.components.append(dataPrefix.get(component).toEscapedString())
        message.command.appName = applicationName

        paramComponent = ProtobufTlv.encode(message)

        requestInterest = Interest(Name(self._controllerName).append("requests").append(paramComponent))

        requestInterest.setInterestLifetimeMilliseconds(4000)
        self._face.makeCommandInterest(requestInterest)
        
        self._face.expressInterest(requestInterest, 
          lambda interest, data : self.onAppRequestData(interest, data, onRequestSuccess, onRequestFailed), 
          lambda interest : self.onAppRequestTimeout(interest, onRequestSuccess, onRequestFailed))
        print "Application publish request sent: " + requestInterest.getName().toUri()
        return

    def onAppRequestData(self, interest, data, onRequestSuccess, onRequestFailed):
        print "Got application publishing request data"
        def onVerified(data):
            responseObj = json.loads(data.getContent().toRawStr())
            if responseObj["status"] == "200":
                if onRequestSuccess:
                    onRequestSuccess()
                else:
                    print "onSetupComplete"
            else:
                print "Verified content: " + data.getContent().toRawStr()
                if onRequestFailed:
                    onRequestFailed(data.getContent().toRawStr())
        def onVerifyFailed(data, reason):
            msg = "Application request response verification failed: " + reason
            print msg
            if onRequestFailed:
                onRequestFailed(msg)

        self._keyChain.verifyData(data, onVerified, onVerifyFailed)
        return

    def onAppRequestTimeout(self, interest, onSetupComplete, onSetupFailed):
        print "Application publishing request times out"
        newInterest = Interest(interest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest,
          lambda interest, data : self.onAppRequestData(interest, data, onSetupComplete, onSetupFailed), 
          lambda interest : self.onAppRequestTimeout(interest, onSetupComplete, onSetupFailed))
        return

###############################################
# Helper functions
###############################################
    def onRegisterFailed(self, prefix):
        print("register failed for prefix " + prefix.getName().toUri())
        return

    def processConfiguration(self, confFile):
        config = BoostInfoParser()
        config.read(confFile)

        # TODO: handle missing configuration, refactor dict representation
        confObj = dict()
        try:
            confObj["identity"] = config["application/identity"][0].value
            confObj["signer"] = config["application/signer"][0].value
        except KeyError as e:
            msg = "Missing key in configuration: " + str(e)
            print msg
            return None
        return confObj

    def getIdentityNameFromCertName(self, certName):
        i = certName.size() - 1

        idString = "KEY"
        while i >= 0:
            if certName.get(i).toEscapedString() == idString:
                break
            i -= 1

        if i < 0:
            print "Error: unexpected certName " + certName.toUri()
            return None

        return Name(certName.getPrefix(i))

#################################
# Getters and setters
#################################
    def getKeyChain(self):
        return self._keyChain

File: framework/ndn_iot_js/js/bootstrap/bootstrap.js
var Bootstrap = function Bootstrap(face)
{
    this.identityManager = new IdentityManager(new IndexedDbIdentityStorage(), new IndexedDbPrivateKeyStorage());
    
    var defaultPolicy = 
      "validator"                + "\n" +
      "{"                        + "\n" +
      "  rule"                   + "\n" +
      "  {"                      + "\n" +
      "    id \"initial rule\""  + "\n" +
      "    for data"             + "\n" +
      "    checker"              + "\n" +
      "    {"                    + "\n" +
      "      type hierarchical"  + "\n" +
      "    }"                    + "\n" +
      "  }"                      + "\n" +
      "}";

    this.policyManager = new ConfigPolicyManager();
    this.policyManager.load(defaultPolicy, "default-policy");

    // keyChain is what we return to the application after successful setup
    // TODO: should we separate keyChain from internal KeyChain used to verify trust schemas?
    this.keyChain = new KeyChain(this.identityManager, this.policyManager)

    this.face = face;
    // setFace for keyChain or else it won't be able to express interests for certs
    this.keyChain.setFace(this.face);
    this.certificateContentCache = new MemoryContentCache(this.face);
    
    this.trustSchemas = {};
}
        
/**
 * Initial keyChain and defaultCertificate setup
 */
Bootstrap.prototype.setupDefaultIdentityAndRoot = function(identityName, signerName, onSetupComplete, onSetupFailed)
{
    var self = this;

    function helper (defaultIdentity) {
        self.identityManager.getDefaultCertificateNameForIdentity(defaultIdentity, function (defaultCertificateName) {
            self.defaultIdentity = defaultIdentity;
            self.defaultCertificateName = defaultCertificateName;

            self.identityManager.getDefaultKeyNameForIdentity(defaultIdentity, function (defaultKeyName) {
                self.defaultKeyName = defaultKeyName;
                // Note we'll not be able to issue face commands before this point
                self.face.setCommandSigningInfo(self.keyChain, self.defaultCertificateName);
                // Serve our own certificate
                self.certificateContentCache.registerPrefix(new Name(self.defaultCertificateName.getPrefix(-1)), function (prefix) {
                    console.log("Prefix registration failed: " + prefix.toUri());
                });
                self.keyChain.getCertificate(self.defaultCertificateName, function (certificate) {
                    self.certificateContentCache.add(certificate);
                    var actualSignerName = certificate.getSignature().getKeyLocator().getKeyName();
                    if (signerName !== undefined && actualSignerName.toUri() !== signerName.toUri()) {
                        var msg = "Configuration signer names mismatch: expected " + signerName.toUri() + "; got " + actualSignerName.toUri();
                        if (onSetupFailed !== undefined)
                            onSetupFailed(msg);
                        return;
                    }
                    self.controllerName = self.getIdentityNameFromCertName(actualSignerName);
                    console.log("Controller name: " + self.controllerName.toUri());
                    
                    self.identityManager.getDefaultCertificateNameForIdentity(self.controllerName, function (certificateName) {
                        self.keyChain.getCertificate(certificateName, function (certificate) {
                            self.controllerCertificate = certificate;   
                            // TODO: this does not seem a good approach, implementation-wise and security implication
                            self.policyManager.certificateCache.insertCertificate(self.controllerCertificate);
                            if (onSetupComplete !== undefined) {
                                onSetupComplete(new Name(self.defaultCertificateName), self.keyChain);
                            }
                        }, function (error) {
                            var msg = "Cannot find controller certificate: " + certificateName.toUri();
                            console.log(error);
                            if (onSetupFailed !== undefined) {
                                onSetupFailed(msg);
                            }
                        })
                    }, function (error) {
                        var msg = "Cannot find default ceritificate for controller identity: " + self.controllerName.toUri();
                        console.log(error);
                        if (onSetupFailed !== undefined) {
                            onSetupFailed(msg);
                        }
                    });
                }, function (error) {
                    var msg = "Certificate does not exist " + self.defaultCertificateName.toUri();
                    if (onSetupFailed !== undefined)
                        onSetupFailed(msg);
                    return;
                });
            }, function (error) {
                console.log(error);
                var msg = "Identity " + defaultIdentity.toUri() + " in configuration does not have a default key. Please configure the device with this identity first";
                if (onSetupFailed !== undefined)
                    onSetupFailed(msg);
                return;
            });
        }, function (error) {
            console.log(error);
            var msg = "Identity " + defaultIdentity.toUri() + " in configuration does not have a default certificate. Please configure the device with this identity first";
            if (onSetupFailed !== undefined)
                onSetupFailed(msg);
            return;
        });
    }
    
    if (identityName === undefined) {
        this.identityManager.getDefaultIdentity(function (defaultIdentityName) {
            helper(defaultIdentityName);
        }, function (error) {
            var msg = "Default identity not configured";
            if (onSetupFailed !== undefined)
                onSetupFailed(msg);
            return;
        });
    } else {
        helper(identityName);
    }
}

/**
 * We don't ask for controller certificate in this implementation, but expects it to be existent per the add device step
 */
/*
    def onControllerCertData(self, interest, data, onSetupComplete, onSetupFailed):
        # TODO: verification rule for received self-signed cert. 
        # So, if a controller comes masquerading in at this point with the right name, it is problematic. Similar with ndn-pi's implementation
        self._controllerCertificate = IdentityCertificate(data)
        # insert root certificate so that we could verify initial trust schemas
        # TODO: this does not seem a good approach, implementation-wise and security implication
        self._keyChain.getPolicyManager()._certificateCache.insertCertificate(self._controllerCertificate)
        try:
            self._identityManager.addCertificate(self._controllerCertificate)
        except SecurityException as e:
            print str(e)
        for schema in self._trustSchemas:
            if "pending-schema" in self._trustSchemas[schema]:
                self._keyChain.verifyData(self._trustSchemas[schema]["pending-schema"], self.onSchemaVerified, self.onSchemaVerificationFailed)
        if onSetupComplete:
            onSetupComplete(Name(self._defaultCertificateName), self._keyChain)
        return

    def onControllerCertTimeout(self, interest, onSetupComplete, onSetupFailed):
        print "Controller certificate interest times out"
        newInterest = Interest(interest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest, 
          lambda interest, data: self.onControllerCertData(interest, data, onSetupComplete, onSetupFailed), 
          lambda interest: self.onControllerCertTimeout(interest, onSetupComplete, onSetupFailed))
        return
*/

/**
 * Handling application consumption (trust schema updates)
 */
// TODO: if trust schema gets over packet size limit, segmentation
Bootstrap.prototype.startTrustSchemaUpdate = function (appPrefix, onUpdateSuccess, onUpdateFailed)
{
    var namespace = appPrefix.toUri()
    if (namespace in this.trustSchemas) {
        if (this.trustSchemas[namespace]["following"] == true) {
            console.log("already following trust schema under this namespace: " + namespace);
            return;
        }
        this.trustSchemas[namespace]["following"] = true;
    } else {
        this.trustSchemas[namespace] = {"following": true, "version": 0, "is-initial": true};
    }

    var initialInterest = new Interest(new Name(namespace).append("_schema"));
    initialInterest.setChildSelector(1);
    var self = this;
    this.face.expressInterest(initialInterest, function (interest, data) {
        self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed);
    }, function (interest) {
        self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed);
    });
    return;
}
        
Bootstrap.prototype.stopTrustSchemaUpdate = function ()
{
    console.log("stopTrustSchemaUpdate not implemented");
    return;
}

Bootstrap.prototype.onSchemaVerified = function (data, onUpdateSuccess, onUpdateFailed)
{
    console.log("trust schema verified: " + data.getName().toUri());
    var version = data.getName().get(-1);
    var namespace = data.getName().getPrefix(-2).toUri();

    if (!(namespace in this.trustSchemas)) {
        console.log("unexpected: received trust schema for application namespace that's not being followed; malformed data name?");
        return;
    }

    if (version.toVersion() <= this.trustSchemas[namespace]["version"]) {
        var msg = "Got out-of-date trust schema";
        console.log(msg);
        if (onUpdateFailed !== undefined) {
            onUpdateFailed(msg);
        }
        return;
    }

    this.trustSchemas[namespace]["version"] = version.toVersion();
    // Remove pending trust schema (while fetching root certificate) logic
    
    var trustSchemaString = data.getContent().toString("binary");
    this.trustSchemas[namespace]["trust-schema"] = trustSchemaString;

    // TODO: what about trust schema for discovery, is discovery its own application?
    var newInterest = new Interest(new Name(data.getName()).getPrefix(-1));
    newInterest.setChildSelector(1);
    var excludeComponent = data.getName().get(-1);
    var exclude = new Exclude();
    exclude.appendAny();
    exclude.appendComponent(version);
    newInterest.setExclude(exclude);
    var self = this;
    this.face.expressInterest(newInterest, function (interest, data) {
        self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed);
    }, function (interest) {
        self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed);
    });

    // Note: this changes the verification rules for root cert, future trust schemas as well; ideally from the outside this doesn't have an impact, but do we want to avoid this?
    // Per reset function in ConfigPolicyManager; For now we don't call reset as we still want root cert in our certCache, instead of asking for it again (when we want to verify) each time we update the trust schema
    // TODO: check if the above note holds for our JS implementation and whether it matters as we are using base64 root by default
    this.policyManager.load(trustSchemaString, "updated-schema");

    if (onUpdateSuccess !== undefined) {
        onUpdateSuccess(trustSchemaString, this.trustSchemas[namespace]["is-initial"]);
    }
    this.trustSchemas[namespace]["is-initial"] = false;
    return;
}

Bootstrap.prototype.onSchemaVerificationFailed = function (data, onUpdateSuccess, onUpdateFailed)
{
    console.log("trust schema verification failed");
    var namespace = data.getName().getPrefix(-2).toUri();
    if (!(namespace in self._trustSchemas)) {
        console.log("unexpected: received trust schema for application namespace that's not being followed; malformed data name?");
        return ;
    }
    
    var newInterest = new Interest(new Name(data.getName()).getPrefix(-1));
    newInterest.setChildSelector(1);
    var excludeComponent = data.getName().get(-1);
    var exclude = new Exclude();
    exclude.appendAny();
    exclude.appendComponent(Name.Component.fromVersion(this.trustSchemas[namespace]["version"]));
    newInterest.setExclude(exclude);

    // Don't immediately ask for potentially the same content again if verification fails
    var self = this;
    setTimeout(4000, function () {
        self.face.expressInterest(newInterest, function (interest, data) {
            self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed);
        }, function (interest) {
            self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed);
        })
    });
    return;
}

Bootstrap.prototype.onTrustSchemaData = function (interest, data, onUpdateSuccess, onUpdateFailed)
{
    console.log("Trust schema received: " + data.getName().toUri());
    var namespace = data.getName().getPrefix(-2).toUri();

    // Process newly received trust schema
    if (this.controllerCertificate === undefined) {
        // We should have controller certificate, don't do pending schema in JS implementation
        console.log("Controller certificate not yet present, verify once it's in place");
    } else {
        // we veriy the received trust schema, should we use an internal KeyChain instead?
        var self = this;
        this.keyChain.verifyData(data, function (data) {
            self.onSchemaVerified(data, onUpdateSuccess, onUpdateFailed);
        }, function (data) {
            self.onSchemaVerificationFailed(data, onUpdateSuccess, onUpdateFailed); 
        });
    }
        
    return;
}

Bootstrap.prototype.onTrustSchemaTimeout = function (interest, onUpdateSuccess, onUpdateFailed)
{
    console.log("Trust schema interest times out: " + interest.getName().toUri());
    var newInterest = new Interest(interest);
    newInterest.refreshNonce();
    var self = this;
    this.face.expressInterest(newInterest, function (interest, data) {
        self.onTrustSchemaData(interest, data, onUpdateSuccess, onUpdateFailed);
    }, function (interest) {
        self.onTrustSchemaTimeout(interest, onUpdateSuccess, onUpdateFailed);
    });
    return
}

/**
 * Handling application producing authorizations
 */
// Wrapper for sendAppRequest, fills in already configured defaultCertificateName
Bootstrap.prototype.requestProducerAuthorization = function (dataPrefix, appName, onRequestSuccess, onRequestFailed)
{
    // TODO: update logic on this part, should the presence of default certificate name be mandatory? 
    // And allow application developer to send app request to a configured root/controller?
    if (this.defaultCertificateName === undefined) {
        console.log("Default certificate is missing! Try setupDefaultIdentityAndRoot first?");
        return;
    }

    this.sendAppRequest(this.defaultCertificateName, dataPrefix, appName, onRequestSuccess, onRequestFailed);
}
        
Bootstrap.prototype.sendAppRequest = function (certificateName, dataPrefix, applicationName, onRequestSuccess, onRequestFailed)
{
    var ProtoBuf = dcodeIO.ProtoBuf;
    var builder = ProtoBuf.loadProtoFile('app-request.proto');
    var descriptor = builder.lookup('AppRequestMessage');
    var AppRequestMessage = descriptor.build();

    var message = new AppRequestMessage();
    message.command = new AppRequestMessage.AppRequest();
    message.command.idName = new AppRequestMessage.Name();
    message.command.dataPrefix = new AppRequestMessage.Name();
    
    for (var i = 0; i < certificateName.size(); i++) {
        message.command.idName.add("components", certificateName.get(i).getValue().buf());
    }
    for (var i = 0; i < dataPrefix.size(); i++) {
        message.command.dataPrefix.add("components", dataPrefix.get(i).getValue().buf());
    }

    message.command.appName = applicationName;

    var paramComponent = new Name.Component(ProtobufTlv.encode(message, descriptor));
    var requestInterest = new Interest
      (new Name(this.controllerName).append("requests").append(paramComponent));
    requestInterest.setInterestLifetimeMilliseconds(4000);
    this.face.makeCommandInterest(requestInterest);
    var self = this;
    this.face.expressInterest(requestInterest, function (interest, data) {
        self.onAppRequestData(interest, data, onRequestSuccess, onRequestFailed); 
    }, function (interest) {
        self.onAppRequestTimeout(interest, onRequestSuccess, onRequestFailed);
    });
    console.log("Application publish request sent: " + requestInterest.getName().toUri())
    return;
}
        
Bootstrap.prototype.onAppRequestData = function (interest, data, onRequestSuccess, onRequestFailed)
{
    console.log("Got application publishing request data");
    function onVerified(data) {
        var responseObj = JSON.parse(data.getContent().toString("binary"));
        if (responseObj["status"] == "200") {
            if (onRequestSuccess !== undefined) {
                onRequestSuccess();
            }
        } else {
            console.log("Verified content: " + data.getContent().toString("binary"));
            if (onRequestFailed !== undefined) {
                onRequestFailed(data.getContent().toString("binary"));
            }
        }
    }
        
    function onVerifyFailed(data) {
        var msg = "Application request response verification failed!"
        console.log(msg);
        if (onRequestFailed !== undefined) {
            onRequestFailed(msg);
        }
    }
        
    this.keyChain.verifyData(data, onVerified, onVerifyFailed);
    return;
}
    
Bootstrap.prototype.onAppRequestTimeout = function(interest, onSetupComplete, onSetupFailed)
{
    console.log("Application publishing request times out");
    var newInterest = new Interest(interest);
    newInterest.refreshNonce();
    var self = this;
    this.face.expressInterest(newInterest, function (interest, data) {
        self.onAppRequestData(interest, data, onSetupComplete, onSetupFailed); 
    }, function (interest) {
        self.onAppRequestTimeout(interest, onSetupComplete, onSetupFailed);
    });
    return;    
}

/**
 * Helper functions
 */
Bootstrap.prototype.onRegisterFailed = function (prefix)
{
    console.log("register failed for prefix " + prefix.getName().toUri());
    return;
}

/*
    def processConfiguration(self, confFile):
        config = BoostInfoParser()
        config.read(confFile)

        # TODO: handle missing configuration, refactor dict representation
        confObj = dict()
        try:
            confObj["identity"] = config["application/identity"][0].value
            confObj["signer"] = config["application/signer"][0].value
        except KeyError as e:
            msg = "Missing key in configuration: " + str(e)
            print msg
            return None
        return confObj
*/

Bootstrap.prototype.getIdentityNameFromCertName = function (certName)
{
    var i = certName.size() - 1;

    var idString = "KEY";
    while (i >= 0) {
        if (certName.get(i).toEscapedString() == idString) {
            break;            
        }
        i -= 1;
    }
        
    if (i < 0) {
        console.log("Error: unexpected certName " + certName.toUri())
        return;
    }
    
    return new Name(certName.getPrefix(i));
}
    
/**
 * Getters and setters
 */
Bootstrap.prototype.getKeyChain = function()
{
    return this.keyChain;
}